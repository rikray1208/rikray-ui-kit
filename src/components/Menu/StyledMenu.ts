import styled, { css } from 'styled-components';

import { setFontSize } from '../../utils/helpers';

const listReset = css`
    margin: 0;
    padding: 0;
    text-indent: 0;
    list-style-type: none;
`;

export const StyledMenu = styled.ul`
    ${listReset};

    display: block;
    width: 100%;
    height: 100%;
`;

export const StyledItem = styled.li<{ isGroup: boolean; heading: boolean }>`
    ${listReset};

    ${({ theme, isGroup, heading }) => css`
        margin: 4px ${theme.Menu['margin-1.5'] + 'rem'};
        cursor: pointer;

        ${setFontSize('Small', 'Menu', { greater: true })};
        color: ${theme.Menu.colorTextSecondary};
        font-weight: ${theme.Menu.fontWeightBold};

        &:not(.ItemGroup) {
            > span:hover {
                color: ${theme.Menu.colorHoverPrimary};
                background: ${theme.Menu.colorShadowPrimary};
            }
        }

        > ul {
            transition: max-height ${theme.Menu.durationSlow + 'ms'} ${theme.Menu.transitionEaseOut} -100ms;
            overflow: hidden;
            max-height: 0;
        }

        &.Active {
            > span {
                color: ${theme.Menu.colorPrimary};
            }

            > ul {
                transition: max-height ${theme.Menu.durationSlow + 'ms'} ${theme.Menu.transitionEaseIn};
                max-height: 999px;
            }
        }

        ${heading &&
        css`
            > span {
                font-weight: ${theme.Menu.fontWeightExtraBold};
                color: ${theme.Menu.colorTextHeading};
            }
        `}

        ${isGroup &&
        css`
            > span {
                color: ${theme.Menu.colorTextTertiary};
            }
        `}
    `}
`;
export const ItemContent = styled.span`
    ${({ theme }) => css`
        display: flex;
        flex-wrap: wrap;
        align-items: center;
        justify-content: space-between;

        padding: 12px;
        border-radius: ${theme.Menu['borderRadius-lg'] + 'px'};

        font: inherit;
    `}
`;

export const DividerUpperStyles = {
    width: '100%',
    marginBottom: 6,
};
