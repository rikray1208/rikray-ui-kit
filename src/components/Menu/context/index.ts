import React, { createContext } from 'react';

import { MenuItem } from '../index';
type State = {
    globalList: MenuItem[];
    activeElements: number[];
};

type Action = { type: 'add'; payload: number[] } | { type: 'deleteItem'; payload: number } | { type: 'clearAll' };

type Dispatch = (action: Action) => void;

export const MenuContext = createContext<{ state: State; dispatch: Dispatch } | undefined>(undefined);
export function MenuReducer(state: State, action: Action) {
    switch (action.type) {
        case 'add': {
            return { ...state, activeElements: [...state.activeElements, ...action.payload] };
        }
        case 'clearAll': {
            return { ...state, activeElements: [] };
        }

        case 'deleteItem': {
            return { ...state, activeElements: state.activeElements.filter((el) => el !== action.payload) };
        }

        default: {
            throw new Error(`Unhandled action type`);
        }
    }
}
