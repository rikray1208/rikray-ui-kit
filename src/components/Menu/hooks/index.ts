import React from 'react';

import { MenuContext } from '../context';

export function useMenu() {
    const context = React.useContext(MenuContext);
    if (context === undefined) {
        throw new Error('useMenu must be used within a MenuProvider');
    }
    return context;
}
