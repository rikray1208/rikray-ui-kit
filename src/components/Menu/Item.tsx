import { ChevronDown, ChevronRight } from 'lucide-react';
import React, { FC } from 'react';

import { MenuList } from './index';

import { useMenu } from './hooks';
import { DividerUpperStyles, ItemContent, StyledItem } from './StyledMenu';

import Divider from '../Divider';

import type { MenuItem } from './index';

const Item: FC<MenuItem> = (props) => {
    const { state } = useMenu();
    const handleClick = (event: React.MouseEvent<HTMLLIElement>) => {
        event.stopPropagation();

        if (props.onClick && !props.group) {
            props.onClick(props);
        }
    };

    const isHeading = !!state.globalList.find((el) => el.id == props.id) || !!props.children;

    return (
        <StyledItem
            onClick={handleClick}
            isGroup={!!props.group}
            heading={isHeading}
            className={props.className + ' ' + (props.group ? 'ItemGroup' : '')}
        >
            <ItemContent>
                {props.divider && <Divider disabled style={DividerUpperStyles} />}
                {props.title}
                {!state.activeElements.includes(props.id) && props.children && <ChevronRight size={16} />}
                {state.activeElements.includes(props.id) && props.children && <ChevronDown size={16} />}
            </ItemContent>

            {props.children && <MenuList list={props.children} />}
        </StyledItem>
    );
};
export default Item;
