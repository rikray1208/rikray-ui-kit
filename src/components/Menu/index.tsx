import React, { FC, useEffect } from 'react';

import MenuProvider from './context/ContextProvider';
import { useMenu } from './hooks';
import Item from './Item';
import { StyledMenu } from './StyledMenu';

import { getMenuPatents } from '../../utils/helpers';

export interface MenuItem {
    id: number;
    title: string;
    group?: boolean;
    divider?: boolean;
    children?: MenuItem[];
    onClick?: (item: MenuItem) => void;
    className?: string;
}

export interface MenuProps {
    list: MenuItem[];
    onSelect?: (item: MenuItem) => void;
    defaultSelectedItem?: number;
}

const Menu: FC<MenuProps> = (props) => {
    const { list, defaultSelectedItem, onSelect } = props;

    return (
        <MenuProvider list={list}>
            <MenuList defaultSelectedItem={defaultSelectedItem} onSelect={onSelect} list={list} />
        </MenuProvider>
    );
};

export const MenuList: FC<MenuProps> = ({ list, onSelect, defaultSelectedItem }) => {
    const { state, dispatch } = useMenu();

    function handleSelect(item: MenuItem) {
        const activeItems = [...getMenuPatents(item.id, state.globalList), item.id];
        if (!state.activeElements.includes(item.id)) {
            dispatch({ type: 'clearAll' });
            dispatch({ type: 'add', payload: activeItems });

            onSelect && onSelect(item);
        }

        if (state.activeElements.includes(item.id) && !!item.children) {
            dispatch({ type: 'deleteItem', payload: item.id });
        }
    }

    useEffect(() => {
        if (defaultSelectedItem) {
            const activeItems = [...getMenuPatents(defaultSelectedItem, state.globalList), defaultSelectedItem];
            dispatch({ type: 'add', payload: activeItems });
        }
    }, [defaultSelectedItem]); // eslint-disable-line

    return (
        <StyledMenu className={'Menu'}>
            {list.map((item) => (
                <Item
                    key={item.id}
                    {...item}
                    className={'MenuItem' + ' ' + (state.activeElements.includes(item.id) ? 'Active' : '')}
                    onClick={handleSelect}
                />
            ))}
        </StyledMenu>
    );
};
export default Menu;
