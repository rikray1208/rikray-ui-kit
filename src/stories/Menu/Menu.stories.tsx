import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import { MenuItem } from '../../components/Menu';
import { Card, Menu, ThemeProvider } from '../../index';

const Items: MenuItem[] = [
    {
        title: 'Getting Started',
        group: true,
        id: 1,
    },
    {
        title: 'Guides',
        id: 2,
        children: [
            {
                title: 'Start new project',
                id: 3,
            },
        ],
    },
    {
        title: 'Components',
        id: 5,
        children: [
            {
                title: 'Navigation',
                id: 6,
                group: true,
                divider: true,
            },
            {
                title: 'Menu',
                id: 7,
            },
            {
                title: 'Step',
                id: 8,
            },
        ],
    },
    {
        title: 'Themes',
        id: 9,
    },
];

export const decorators = [
    (Story: any) => (
        <ThemeProvider>
            <Card>
                <aside
                    style={{
                        width: 400,
                    }}
                >
                    <Story />
                </aside>
            </Card>
        </ThemeProvider>
    ),
];

export default {
    title: 'Design System/Navigation/Menu',
    component: Menu,
    decorators: decorators,
} as ComponentMeta<typeof Menu>;

const Template: ComponentStory<typeof Menu> = (args) => <Menu {...args} />;

export const DefaultMenu = Template.bind({});
DefaultMenu.args = {
    list: Items,
    defaultSelectedItem: 3,
    onSelect: (item) => console.log('Selected item: ' + item.id),
};
