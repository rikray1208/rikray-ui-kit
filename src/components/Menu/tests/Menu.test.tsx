import React from 'react';
import { create } from 'react-test-renderer';

import { ThemeProvider } from '../../../index';
import Menu from '../../Menu';

const MockItems = [
    {
        title: 'Getting Started',
        group: true,
        id: 1,
    },
    {
        title: 'Guides',
        id: 2,
        children: [
            {
                title: 'Start new project',
                id: 3,
            },
        ],
    },
    {
        title: 'Components',
        id: 5,
        children: [
            {
                title: 'Navigation',
                id: 6,
                group: true,
                divider: true,
            },
            {
                title: 'Menu',
                id: 7,
            },
            {
                title: 'Step',
                id: 8,
            },
        ],
    },
    {
        title: 'Themes',
        id: 9,
    },
];

describe('Menu', () => {
    it('should return a Menu', () => {
        const element = (
            <ThemeProvider>
                <Menu list={MockItems} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a Menu with default selected item', () => {
        const element = (
            <ThemeProvider>
                <Menu list={MockItems} defaultSelectedItem={3} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
});
