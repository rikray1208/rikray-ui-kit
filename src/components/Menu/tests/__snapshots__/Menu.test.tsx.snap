// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should return a Menu 1`] = `
<ul
  className="sc-hhOBVt dPKplo Menu"
>
  <li
    className="sc-gYbzsP imrgbM MenuItem  ItemGroup"
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Getting Started
    </span>
  </li>
  <li
    className="sc-gYbzsP iiYEgl MenuItem  "
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Guides
      <svg
        className="lucide lucide-chevron-right"
        fill="none"
        height={16}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </span>
    <ul
      className="sc-hhOBVt dPKplo Menu"
    >
      <li
        className="sc-gYbzsP cxSJwU MenuItem  "
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          Start new project
        </span>
      </li>
    </ul>
  </li>
  <li
    className="sc-gYbzsP iiYEgl MenuItem  "
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Components
      <svg
        className="lucide lucide-chevron-right"
        fill="none"
        height={16}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </span>
    <ul
      className="sc-hhOBVt dPKplo Menu"
    >
      <li
        className="sc-gYbzsP kMqVUh MenuItem  ItemGroup"
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          <div
            className="sc-kMjNwy eedxQV"
            style={
              {
                "marginBottom": 6,
                "width": "100%",
              }
            }
          >
            <div
              className="sc-ilhmMj jRudac"
              disabled={true}
            />
            <div
              className="sc-ilhmMj jRudac"
              disabled={true}
            />
          </div>
          Navigation
        </span>
      </li>
      <li
        className="sc-gYbzsP cxSJwU MenuItem  "
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          Menu
        </span>
      </li>
      <li
        className="sc-gYbzsP cxSJwU MenuItem  "
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          Step
        </span>
      </li>
    </ul>
  </li>
  <li
    className="sc-gYbzsP iiYEgl MenuItem  "
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Themes
    </span>
  </li>
</ul>
`;

exports[`Menu should return a Menu with default selected item 1`] = `
<ul
  className="sc-hhOBVt dPKplo Menu"
>
  <li
    className="sc-gYbzsP imrgbM MenuItem  ItemGroup"
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Getting Started
    </span>
  </li>
  <li
    className="sc-gYbzsP iiYEgl MenuItem  "
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Guides
      <svg
        className="lucide lucide-chevron-right"
        fill="none"
        height={16}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </span>
    <ul
      className="sc-hhOBVt dPKplo Menu"
    >
      <li
        className="sc-gYbzsP cxSJwU MenuItem  "
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          Start new project
        </span>
      </li>
    </ul>
  </li>
  <li
    className="sc-gYbzsP iiYEgl MenuItem  "
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Components
      <svg
        className="lucide lucide-chevron-right"
        fill="none"
        height={16}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </span>
    <ul
      className="sc-hhOBVt dPKplo Menu"
    >
      <li
        className="sc-gYbzsP kMqVUh MenuItem  ItemGroup"
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          <div
            className="sc-kMjNwy eedxQV"
            style={
              {
                "marginBottom": 6,
                "width": "100%",
              }
            }
          >
            <div
              className="sc-ilhmMj jRudac"
              disabled={true}
            />
            <div
              className="sc-ilhmMj jRudac"
              disabled={true}
            />
          </div>
          Navigation
        </span>
      </li>
      <li
        className="sc-gYbzsP cxSJwU MenuItem  "
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          Menu
        </span>
      </li>
      <li
        className="sc-gYbzsP cxSJwU MenuItem  "
        onClick={[Function]}
      >
        <span
          className="sc-cCjUiG dehrLs"
        >
          Step
        </span>
      </li>
    </ul>
  </li>
  <li
    className="sc-gYbzsP iiYEgl MenuItem  "
    onClick={[Function]}
  >
    <span
      className="sc-cCjUiG dehrLs"
    >
      Themes
    </span>
  </li>
</ul>
`;
